kimchi/src/linearization.rs

```diff
@@ -47,6 +52,10 @@ pub fn constraints_expr<F: FftField + SquareRootField>(
         expr += ChaChaFinal::combined_constraints(&powers_of_alpha);
     }
 
+    if foreign_mul {
+        expr += foreign_mul::combined_constraints(&powers_of_alpha);
```

Comment:
Chacha was actually a single call like this initially, but I felt like it was too much magic and not really showing us the cost of having the gate. Now we can clearly see that ChaCha has 4 gates/selector polynomials, whereas here it's not clear what the cost is. wdyt?

Reply:
I'm thinking of the gate (aka gadget) as a collection of `CircuitGate`.

I think a gate (i.e. foreign_mul, chacha, cairo) is atomic.  That is, the caller must not reach in and use only part of it, such as individual rows/CircuitGates.  They are designed to be used all together, not individually.

Therefore, I think it makes sense to encapsulate the internal details, such as how many CircuitGates is it comprised of, so that the caller doesn't need to concern itself with these concerns.

Following this line of reasoning, I don't think the above code should need to manually build the gate from it's individual CircuitGates each time it's called.  Instead, I feel like the caller should go "Oi -- give us the foreign field multiplication combined constraints"

I think counting or tracking stats how many constraints are added should be done elsewhere and just commented if it is relevant for performance reasons.